### ISSUE TITLE
Proxy support

### ISSUE THREAD

**Issue by CaptainMorgs** [PaulMcInnis/JobFunnel#44]
Proxy support would be nice if you were somewhere that requires them.

E.g. in indeed.py

` http_proxy  = "http.foo.com:8000"
        https_proxy = "http.foo.com:8000"
        proxyDict = { 
              "http"  : http_proxy, 
              "https" : https_proxy
            }       
request_HTML = get(search, headers=self.headers, proxies=proxyDict)`

**Comment by PaulMcInnis** [comment]
Agreed, 

I’ll tag this as a feature request, feel free to take it on.

**Comment by rodhod** [comment]
@CaptainMorgs @PaulMcInnis 
Is it as simple as just adding the "https_proxy" and "http_proxy" variables and a dictionary to the "indeed.py", "monster.py", "glassdoor.py" files? If so, I see multiple "get()" methods in "indeed.py" for example and one variable in the scrape function called "request_HTML", is it simply a matter of editing this function? 
I have some experience with python and I would like implement this proxy functionality if someone is not already working on it. Is it possible for someone to put me on the right path with this enhancement? It is my first issue/contribution to someone else's proejct.

**Comment by PaulMcInnis** [comment]
Go for it, ill assign it to you :) open a PR when you are able.

**Comment by rodhod** [comment]
thank you @PaulMcInnis 
i need a little clarification though, what I'm thinking is:
1. define `http_proxy` and `https_proxy` as global variables to all three `job_site.py` scripts
2. define a `proxyDict{}` in the global scope of all three `job_site.py` scripts
3. modify the `request_HTML` variable definition inside the `scrape()` function to use proxies e.g. (`get(search, headers=self.headers, proxies=proxyDict`)
4. add proxy config to the settings yaml to set the definitions of variables `http_proxy` and `https_proxy`

then i'm guessing we have the option of asking the user for the proxy info in the CLI or asking them to set it themselves in the settings yaml. how does all that sound?

sorry for so many questions, just trying to get familiar with the codebase.

**Comment by PaulMcInnis** [comment]
Good to ask questions! 

I think the proxy configuration should be an attribute of the base JobFunnel class, which all the jobsite.py providers inherit from. 

We could have an attribute of JobFunnel provide http/s proxy configuration and then we can handle it in the request formatting for each provider.

**Comment by rodhod** [comment]
@PaulMcInnis you are right i just realized i said it wrong. all those function are within that JobFunnel class including `scrape()`. 

i think i can figure it out let me tinker with it a little bit. 
i am really wondering how the settings yaml get its global config values to the `job_site.py` scripts. i know it goes through the parser in the config folder but how do the scripts import these values or how do they "see" them?

**Comment by PaulMcInnis** [comment]
Well since the other scripts are sub classes of the abstract base class “JobFunnel” they have access to the attributes of that base class, which is initialized with the YAML :)

**Comment by rodhod** [comment]
@PaulMcInnis `get()` is used in some of the functions prior to `scrape()` should i edit all the instances of `get()` or just the `request_HTML=get()` instance within the `scrape()` function?

**Comment by markkvdb** [comment]
You’d have to change all calls to the get function. A solution could be to define a new ‘get’ function in the parent class ‘JobFunnel’. This could for example be done using partial functions where you’d set the ‘proxies’ argument to your specification. 

You’d then have to call this new function ‘get’ which is a member of the ‘JobFunnel’ objects (also derived objects). Note that this would also require you to set the proxies to the None option in case someone doesn’t use proxies. 

I believe this is an elegant way to implement proxies without changing the (internal) interface too much. 

Let me know what you think!

**Comment by rodhod** [comment]
so this is where i am so far:

1. I added this to the `setttings.yaml` config
```
# Proxy Configuration (Please specify address and port)
proxy_config:
	# example '0.0.0.0'
	address: None
	# example '0000'
	port: None
```
2. i added these lines to the bottom of the `parse_config()` functions in `parser.py`
```
#proxy config
    config['proxy_config']['address'] = default_yaml['proxy_config']['address']
    config['proxy_config']['port'] = default_yaml['proxy_config']['port']
```
3. I added these lines to the `Indeed(JobFunnel)` class in `indeed.py`
```
config = parse_config()
    
http_proxy = 'http://' + config['proxy_config']['address'] + ':' + config['proxy_config']['port']
https_proxy = 'http://' + config['proxy_config']
['address'] + ':' + config['proxy_config']['port']

proxies = {
    'http': http_proxy,
     https': https_proxy,
}
```
also added this to the head of `indeed.py` to make the above code work
```
from .config.parser import parse_config
```

**Comment by rodhod** [comment]
what do you think? 
if that works i can implement it using your above recommendation with a new 'get' function or I can just change all the 'get' calls.

**Comment by PaulMcInnis** [comment]
@rodhod I recommend checking out the other place parse_config is called, you should set something like self.proxy_url in JobFunnel from args and then access it in indeed/etc the same way(self.proxy_url)  ;)

**Comment by rodhod** [comment]
@PaulMcInnis perfect! i knew there was a way access the args without doing the config thing. but tell me this how do the `args['setting-name']` get info from the config or parse config? maybe i don't understand how `args[]` work. parse_config is called in main and i dont see how that data gets from `config[]` to `self.*`

nevertheless ill try implementing it properly with `self.proxy_url`

**Comment by markkvdb** [comment]
Good job! I do have a few suggestions.

1. Comment out the proxy config when it is not used. In the parse_config you would check if the `proxy_config` is set. 

2. Here you could add an if statement as

```python
config['proxy_config'] = default_yaml['proxy_config']
if given_yaml_path is not None:
    config['proxy_config'] = given_yaml['proxy_config']
if cli.set_delay is not None:
    config['proxy_config'] = cli.proxy_config
```

Note that config['proxy_config`] can be None if it is not set in any of these options. Furthermore, you have to add support for the proxy config in the argument settings.

3. What I would do instead is to set the `get` function in the __init__ class of the `JobFunnel` class. This could be something like

```python
if args['proxy_config'] is not None:
    self.proxies = {
         'http_proxy': 'http://' + args['proxy_config']['address'] + ':' + args['proxy_config']['port'],
         'https_proxy': 'https://' + args['proxy_config']['address'] + ':' + args['proxy_config']['port']
    }
else:
    self.proxies = None
``` 

and later (if you need to give a dict with http and https proxies:

```python
if self.proxy_config is not None:
    self.get = partial(requests.get, proxy=proxies)
```

**Comment by markkvdb** [comment]
Of course you then have to replace all the `get` calls in the `JobFunnel` and child classes with a matching call to `self.get`. When `self.proxies` is set to `None` then nothing changes and if a proxy is set then the proxy will always be used for the get calls. I hope this helps.

**Comment by bunsenmurder** [comment]
@markkvdb you should take a look into [session objects](https://requests.readthedocs.io/en/master/user/advanced/), you can do some real cool stuff with them and set the proxy without needing to import partial.

**Comment by rodhod** [comment]
okay @bunsenmurder that's makes this much easier because we can use session to set a proxy for the lifetime of a session object. something like:
```
self.proxies = {
         'http_proxy': 'http://' + args['proxy_config']['address'] + ':' + args['proxy_config']['port'],
         'https_proxy': 'https://' + args['proxy_config']['address'] + ':' + args['proxy_config']['port']
    }
s = requests.session()
s.proxies.update(self.proxies)
s.get(search, headers=self.headers)
```
now every call of `get()`has a proxy attached to the session object; i.e. no need to change all calls.
I wonder if I can do this in the `__init__` function in `indeed.py` and etc? or should I do it just outside in the JobFunnel global class? @PaulMcInnis @markkvdb

**Comment by markkvdb** [comment]
I have taken @bunsenmurder advice and incorporated the session objects in my (merged) pull request for proxy support #51. I think this issue can therefore be closed. @PaulMcInnis you will have the final call on this.

**Comment by PaulMcInnis** [comment]
Closing
